site/source/docs/api_reference/vr.h.rst

```diff
@@ -0,0 +1,378 @@
+	:rtype: int
+
+.. c:function:: int emscripten_vr_version_major()
+
+	Major version of the WebVR API currently supported by the browser.
+
+	:returns: major version of WebVR, or `-1` if not supported or API not initialized.
+	:rtype: int
+
+Display Functions
+=================
+
+All of the following functions require :c:func:`emscripten_vr_init` to have been
+called the return value of :c:func:`emscripten_vr_ready` to be `true`.
+
+.. c:function:: int emscripten_vr_count_displays()
+
+	:returns: Number of displays connected.
+	:rtype: int
+
+.. c:function:: VRDisplayHandle emscripten_vr_get_display_handle(int displayIndex)
+
+	:param int displayIndex: index of display (inclusive 0 to exclusive :c:func:`emscripten_vr_count_displays`).
+	:returns: handle for a VR display.
+	:rtype: VRDisplayHandle
+
+.. c:function:: char* emscripten_vr_get_display_name(VRDisplayHandle handle)
+
+	Get a user-readable name which identifies the VR display.
+
+	:param VRDisplayHandle handle: |display-handle-parameter-doc|
+	:returns: name of the VR display or `0 (NULL)` if the handle is invalid.
+	:rtype: char*
+
+.. c:function:: bool emscripten_vr_get_display_connected(VRDisplayHandle handle)
+
+	:param VRDisplayHandle handle: |display-handle-parameter-doc|
+	:returns: `true` if the display is connected, `false` otherwise or when
+		the handle is invalid.
+	:rtype: bool
+
+.. c:function:: bool emscripten_vr_get_display_presenting(VRDisplayHandle handle)
+
+	See also :c:func:`emscripten_vr_request_present`.
+
+	:param VRDisplayHandle handle: |display-handle-parameter-doc|
+	:returns: `true` if the display is currently presenting, `false` otherwise
+		or when the handle is invalid.
+	:rtype: bool
+
+.. c:function:: int emscripten_vr_get_display_capabilities(VRDisplayHandle handle, VRDisplayCapabilities* displayCaps)
+
+	:param VRDisplayHandle handle: |display-handle-parameter-doc|
+	:param VRDisplayCapabilities displayCaps: receives capabilities of the VR display.
+	:returns: |display-function-return-doc|
+	:rtype: bool
+
+.. c:function:: int emscripten_vr_get_eye_parameters(VRDisplayHandle handle, VREye whichEye, VREyeParameters* eyeParams)
+
+	:param VRDisplayHandle handle: |display-handle-parameter-doc|
+	:param VREye whichEye: which eye to query parameters for.
+	:param VREyeParameters eyeParam: receives the parameters for requested eye.
+	:returns: |display-function-return-doc|
+	:rtype: bool
+
+Render Loop
+===========
+
+In contrast to the usual emscripten main loop (see :ref:`emscripten-h-browser-execution-environment`),
+VR displays require their own rendering loop which is independent from the main loop. The rendering
+loop can be set per display and will act like a main loop with timing mode ``EM_TIMING_RAF`` until the
+display is requested to present, as of which it will run at the VR display's refresh rate.
+
+.. c:function:: void emscripten_vr_set_display_render_loop(VRDisplayHandle handle, em_vr_callback_func callback)
```

Comment:
Here there's also a question about what is the thread context where the render loop is called - probably we want to have the same as the multithreaded emscripten set main loop, which runs the render loop in the context of the calling thread. Then there's a question of what if a vr render loop is registered multiple times in one thread for one device, and what happens if a multiple vr render loops for the same device are registered in separate threads.

Reply:
> probably we want to have the same as the multithreaded emscripten set main loop

I have not used multithreading in Javascript before, so I don't know the consequences of one or the other. If calling WebGL is only allowed from the main loop, then that is obviously what we want, yes :)

> what if a vr render loop is registered multiple times in one thread for one device

It is asserted that there can only be one renderloop per display. The following `.. tip::` section clarifies that there can only be one. Or what do you mean?

> multiple vr render loops for the same device are registered in separate threads

I would expect the memory of WebVR module to be shared between the devices and therefore the same assertion would fail if a render loop is already set.